flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/util/RexProgramExtractor.scala

```diff
@@ -0,0 +1,184 @@
+          }
+        })
+        (convertedExpressions.toArray, unconvertedRexNodes.toArray)
+
+      case _ => (Array.empty, Array.empty)
+    }
+  }
+}
+
+/**
+  * An RexVisitor to extract all referenced input fields
+  */
+class InputRefVisitor extends RexVisitorImpl[Unit](true) {
+
+  private var fields = mutable.LinkedHashSet[Int]()
+
+  def getFields: Array[Int] = fields.toArray
+
+  override def visitInputRef(inputRef: RexInputRef): Unit =
+    fields += inputRef.getIndex
+
+  override def visitCall(call: RexCall): Unit =
+    call.operands.foreach(operand => operand.accept(this))
+}
+
+/**
+  * An RexVisitor to convert RexNode to Expression.
+  *
+  * @param inputNames      The input names of the relation node
+  * @param functionCatalog The function catalog
+  */
+class ConvertToExpression(
+    inputNames: Array[String],
+    functionCatalog: FunctionCatalog)
+    extends RexVisitor[Option[Expression]] {
+
+  override def visitInputRef(inputRef: RexInputRef): Option[Expression] = {
+    Preconditions.checkArgument(inputRef.getIndex < inputNames.length)
+    Option(ResolvedFieldReference(
+      inputNames(inputRef.getIndex),
+      FlinkTypeFactory.toTypeInfo(inputRef.getType)
+    ))
+  }
+
+  override def visitLocalRef(localRef: RexLocalRef): Option[Expression] = {
+    throw new TableException("Bug: RefLocalRef should have been expanded")
+  }
+
+  override def visitLiteral(literal: RexLiteral): Option[Expression] = {
+    Option(Literal(literal.getValue, FlinkTypeFactory.toTypeInfo(literal.getType)))
+  }
+
+  override def visitCall(call: RexCall): Option[Expression] = {
+    val operands = call.getOperands.map(
+      operand => operand.accept(this).orNull
+    )
+
+    // return null if we cannot translate all the operands of the call
+    if (operands.contains(null)) {
+      Option.empty
+    } else {
+        call.getOperator match {
+          case function: SqlFunction =>
+            lookupFunction(replace(function.getName), operands)
+          case postfix: SqlPostfixOperator =>
+            lookupFunction(replace(postfix.getName), operands)
+          case operator@_ =>
+            lookupFunction(replace(s"${operator.getKind}"), operands)
+      }
+    }
+  }
+
+  override def visitFieldAccess(fieldAccess: RexFieldAccess): Option[Expression] = Option.empty
+
+  override def visitCorrelVariable(correlVariable: RexCorrelVariable): Option[Expression] =
+    Option.empty
+
+  override def visitRangeRef(rangeRef: RexRangeRef): Option[Expression] = Option.empty
+
+  override def visitSubQuery(subQuery: RexSubQuery): Option[Expression] = Option.empty
+
+  override def visitDynamicParam(dynamicParam: RexDynamicParam): Option[Expression] = Option.empty
+
+  override def visitOver(over: RexOver): Option[Expression] = Option.empty
+
+  private def lookupFunction(name: String, operands: Seq[Expression]): Option[Expression] = {
+    Try(functionCatalog.lookupFunction(name, operands)) match {
```

Comment:
Yes, i've changed the passed in functionCatalog to an instance which only holds all built-in functions.